File: tehreer-android/src/main/java/com/mta/tehreer/internal/layout/TextRun.java
/*
 * Copyright (C) 2019-2021 Muhammad Tayyab Akram
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *      http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

package com.mta.tehreer.internal.layout;

import android.graphics.Canvas;
import android.graphics.RectF;

import androidx.annotation.NonNull;
import androidx.annotation.Size;

import com.mta.tehreer.collections.FloatList;
import com.mta.tehreer.collections.IntList;
import com.mta.tehreer.collections.PointList;
import com.mta.tehreer.graphics.Renderer;
import com.mta.tehreer.graphics.Typeface;
import com.mta.tehreer.sfnt.WritingDirection;

import java.util.List;

public interface TextRun {
    int getCharStart();
    int getCharEnd();
    boolean isBackward();
    byte getBidiLevel();
    @NonNull List<Object> getSpans();

    int getStartExtraLength();
    int getEndExtraLength();

    @NonNull Typeface getTypeface();
    float getTypeSize();
    @NonNull WritingDirection getWritingDirection();

    int getGlyphCount();
    @NonNull IntList getGlyphIds();
    @NonNull PointList getGlyphOffsets();
    @NonNull FloatList getGlyphAdvances();

    @NonNull IntList getClusterMap();
    @NonNull FloatList getCaretEdges();

    float getAscent();
    float getDescent();
    float getLeading();

    float getWidth();
    float getHeight();

    int getClusterStart(int charIndex);
    int getClusterEnd(int charIndex);

    @NonNull @Size(2) int[] getGlyphRangeForChars(int fromIndex, int toIndex);
    int getLeadingGlyphIndex(int charIndex);
    int getTrailingGlyphIndex(int charIndex);

    float getCaretBoundary(int fromIndex, int toIndex);
    float getCaretEdge(int charIndex);

    float getRangeDistance(int fromIndex, int toIndex);
    int computeNearestCharIndex(float distance);

    float computeTypographicExtent(int glyphStart, int glyphEnd);
    @NonNull RectF computeBoundingBox(@NonNull Renderer renderer, int glyphStart, int glyphEnd);

    void draw(@NonNull Renderer renderer, @NonNull Canvas canvas);
}

File: tehreer-android/src/main/java/com/mta/tehreer/internal/layout/TextRun.kt
/*
 * Copyright (C) 2023 Muhammad Tayyab Akram
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *      http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

package com.mta.tehreer.internal.layout

import android.graphics.Canvas
import com.mta.tehreer.sfnt.WritingDirection
import com.mta.tehreer.collections.IntList
import com.mta.tehreer.collections.PointList
import com.mta.tehreer.collections.FloatList
import android.graphics.RectF
import com.mta.tehreer.graphics.Renderer
import com.mta.tehreer.graphics.Typeface

internal interface TextRun {
    val charStart: Int
    val charEnd: Int
    val isBackward: Boolean
    val bidiLevel: Byte
    val spans: List<Any?>

    val startExtraLength: Int
    val endExtraLength: Int

    val typeface: Typeface
    val typeSize: Float
    val writingDirection: WritingDirection

    val glyphCount: Int
    val glyphIds: IntList
    val glyphOffsets: PointList
    val glyphAdvances: FloatList

    val clusterMap: IntList
    val caretEdges: FloatList

    val ascent: Float
    val descent: Float
    val leading: Float

    val width: Float
    val height: Float

    fun getClusterStart(charIndex: Int): Int
    fun getClusterEnd(charIndex: Int): Int

    fun getGlyphRangeForChars(fromIndex: Int, toIndex: Int): IntArray
    fun getLeadingGlyphIndex(charIndex: Int): Int
    fun getTrailingGlyphIndex(charIndex: Int): Int

    fun getCaretBoundary(fromIndex: Int, toIndex: Int): Float
    fun getCaretEdge(charIndex: Int): Float

    fun getRangeDistance(fromIndex: Int, toIndex: Int): Float
    fun computeNearestCharIndex(distance: Float): Int

    fun computeTypographicExtent(glyphStart: Int, glyphEnd: Int): Float
    fun computeBoundingBox(renderer: Renderer, glyphStart: Int, glyphEnd: Int): RectF

    fun draw(renderer: Renderer, canvas: Canvas)
}
